x/dsmr/node_test.go

```diff
@@ -994,11 +1057,218 @@ func TestNode_NewBlock_IncludesChunkCerts(t *testing.T) {
+	r.NoError(node.Accept(context.Background(), blk))
+
+	r.NoError(node.storage.AddLocalChunkWithCert(anotherChunk, &anotherChunkCert))
+	_, err = node.BuildBlock(context.Background(), blk, 3)
+	r.NoError(err)
+}
+
+func TestNode_Verify_Chunks(t *testing.T) {
+	r := require.New(t)
+
+	tests := []struct {
+		name                       string
+		parentBlocks               [][]int // for each parent, a list of the chunks included.
+		chunks                     []int
+		timestamp                  int64
+		verifyWantErr              error
+		buildWantErr               error
+		validalidityWindowDuration time.Duration
+	}{
+		{
+			name:          "three blocks, unique chunks",
+			parentBlocks:  [][]int{{1}, {2}},
+			chunks:        []int{4},
+			timestamp:     4,
+			verifyWantErr: nil,
+		},
+		{
+			name:          "two blocks one duplicate chunk",
+			parentBlocks:  [][]int{{0, 2}},
+			chunks:        []int{2, 4},
+			timestamp:     3,
+			verifyWantErr: validitywindow.ErrDuplicateContainer,
+			buildWantErr:  nil, // build would filter out duplicate chunks, hence no error.
+		},
+		{
+			name:          "one block duplicate chunks",
+			parentBlocks:  [][]int{},
+			chunks:        []int{2, 2},
+			timestamp:     2,
+			verifyWantErr: validitywindow.ErrDuplicateContainer,
+			buildWantErr:  nil, // build would filter out duplicate chunks, hence no error.
+		},
+		{
+			name:          "three blocks non consecutive duplicate chunks",
+			parentBlocks:  [][]int{{3}, {2}},
+			chunks:        []int{3},
+			timestamp:     4,
+			verifyWantErr: validitywindow.ErrDuplicateContainer,
+			buildWantErr:  ErrNoAvailableChunkCerts,
+		},
+		{
+			name:          "monotonic timestamping",
+			parentBlocks:  [][]int{},
+			chunks:        []int{1},
+			timestamp:     0,
+			verifyWantErr: ErrInvalidBlockHeight,
+			buildWantErr:  ErrTimestampNotMonotonicallyIncreasing,
+		},
+		{
+			name:          "empty block",
+			parentBlocks:  [][]int{{1}, {2}, {3}, {4}},
+			chunks:        []int{},
+			timestamp:     6,
+			verifyWantErr: ErrEmptyBlock,
+			buildWantErr:  ErrNoAvailableChunkCerts,
+		},
+		{
+			name:                       "three blocks non consecutive duplicate chunks outside validity window",
+			parentBlocks:               [][]int{{4}, {5}},
+			chunks:                     []int{4},
+			timestamp:                  4,
+			verifyWantErr:              validitywindow.ErrDuplicateContainer, // this isn't ideal, since it would disqualify a duplicate chunk even when it's outside the validity window. However, it default to the correct direction.
+			buildWantErr:               ErrNoAvailableChunkCerts,
+			validalidityWindowDuration: 1,
+		},
+	}
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			validationWindow := testCase.validalidityWindowDuration
+			if validationWindow == 0 {
+				validationWindow = testingDefaultValidityWindowDuration
+			}
+
+			var node *Node[tx]
+			nodes, indexer := newNodes(t, 1)
+			node = nodes[0]
```

Comment:
I'm not sure the validity window should be separate from the nodes, since each node maintains its own index

Reply:
it's not the validity window itself, but rather the validity window duration. The index remains the same.

Reply:
I *could* create an indexer per node, if you think it would be better.. but it would be a bit of over engineering solution given that it's being used to retrieve previously accepted blocks. Please let me know if you think it's the right move and I'll make it so.